core/commands/keystore.go

```diff
@@ -285,3 +290,108 @@ func keyOutputListEncoders() cmds.EncoderFunc {
 		return nil
 	})
 }
+
+const (
+	keyStoreOldKeyOptionName    = "oldkey"
+	keyStoreAlgorithmDefault    = options.RSAKey
+	keyStoreAlgorithmOptionName = "algorithm"
+	keyStoreBitsOptionName      = "bits"
+)
+
+var keyRotateCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Rotates the ipfs identity.",
+		ShortDescription: `
+Generates a new ipfs identity and saves it to the ipfs config file.
+The daemon must not be running when calling this command.
+
+ipfs uses a repository in the local file system. By default, the repo is
+located at ~/.ipfs. To change the repo location, set the $IPFS_PATH
+environment variable:
+
+    export IPFS_PATH=/path/to/ipfsrepo
+`,
+	},
+	Arguments: []cmds.Argument{},
+	Options: []cmds.Option{
+		cmds.StringOption(keyStoreOldKeyOptionName, "o", "Keystore name for the old/rotated-out key."),
+		cmds.StringOption(keyStoreAlgorithmOptionName, "a", "Cryptographic algorithm to use for key generation.").WithDefault(keyStoreAlgorithmDefault),
+		cmds.IntOption(keyStoreBitsOptionName, "b", "Number of bits to use in the generated RSA private key."),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		cctx := env.(*oldcmds.Context)
+		daemonLocked, err := fsrepo.LockedByOtherProcess(cctx.ConfigRoot)
+		if err != nil {
+			return err
+		}
+
+		log.Info("checking if daemon is running...")
+		if daemonLocked {
+			log.Debug("ipfs daemon is running")
+			e := "ipfs daemon is running. please stop it to run this command"
+			return cmds.ClientError(e)
+		}
+
+		return nil
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		cctx := env.(*oldcmds.Context)
+		nBitsForKeypair, nBitsGiven := req.Options[keyStoreBitsOptionName].(int)
+		algorithm, _ := req.Options[keyStoreAlgorithmOptionName].(string)
+		oldKey, ok := req.Options[keyStoreOldKeyOptionName].(string)
+		if !ok {
+			return fmt.Errorf("keystore name for backing up old key must be provided")
+		}
+		return doRotate(os.Stdout, cctx.ConfigRoot, oldKey, algorithm, nBitsForKeypair, nBitsGiven)
```

Comment:
Hm. This shouldn't use stdout. We should probably:

1. Create a pipe.
2. Emit the reader.
2. Pass the writer to doRotate.

Reply:
On the other hand, it would be _nice_ to output structured information (e.g., the new peer ID). Maybe we should just:

1. Create a pipe.
2. Pass the writer to doRotate.
3. For every line written, emit `Result{Message: ...}`.
4. At the end, emit `Result{ID: ...}`?

But this isn't _really_ necessary as the user can always just query `ipfs id`.